Cannot call generic function with multiple types

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, "Arch Linux"
Processor: 12 cpus, 64bit, little endian, AMD Ryzen 5 5600X 6-Core Processor
CC version: cc (GCC) 11.1.0

getwd: /tmp
vmodules: /home/rcorre/.vmodules
vroot: /home/rcorre/src/v
vexe: /home/rcorre/src/v/v
vexe mtime: 2021-05-24 01:13:01
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.2 f1174da.b9c8649

Git version: git version 2.31.1
Git vroot status: weekly.2021.20-58-gb9c86493
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 7543de81
```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg example.v`
```v
struct Response<T> {
	result T
}

pub fn send<T>(res T) ? {
	msg := Response<T>{
		result: res
	}
	println('$msg')
}

struct Foo {}

struct Bar {}

fn main() {
	send(Foo{}) ?
	send(Bar{}) ?
}

```

**What did you expect to see?**

You can call a generic function with multiple types


**What did you see instead?**
```
==================
/tmp/v/example.13433723120546161468.tmp.c:9190: error: 'main__Response_T_main__Bar' undeclared
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

If you were not working with C interop, this is a compiler bug, please report the bug using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -3238,7 +3238,7 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
 		}
 		return
 	}
-	//
+
 	is_decl := assign_stmt.op == .decl_assign
 	for i, left in assign_stmt.left {
 		if left is ast.CallExpr {
@@ -3280,6 +3280,13 @@ pub fn (mut c Checker) assign_stmt(mut assign_stmt ast.AssignStmt) {
 		right := if i < assign_stmt.right.len { assign_stmt.right[i] } else { assign_stmt.right[0] }
 		mut right_type := assign_stmt.right_types[i]
 		if is_decl {
+			// check generic struct init and return unwrap generic struct type
+			if right is ast.StructInit {
+				if right.typ.has_flag(.generic) {
+					c.expr(right)
+					right_type = right.typ
+				}
+			}
 			if right.is_auto_deref_var() {
 				left_type = c.table.mktyp(right_type.deref())
 			} else {

--- a/vlib/v/tests/generics_multi_types_struct_init_test.v
+++ b/vlib/v/tests/generics_multi_types_struct_init_test.v
@@ -1,0 +1,41 @@
+struct Response<T> {
+	result T
+}
+
+pub fn send<T>(res T) string {
+	msg := Response<T>{
+		result: res
+	}
+	return '$msg'
+}
+
+struct Foo {}
+
+struct Bar {}
+
+fn test_generics_multi_types_struct_init() {
+	mut ret := send(Foo{})
+	println(ret)
+	assert ret.contains('Response<Foo>{')
+	assert ret.contains('result: Foo{}')
+
+	ret = send(Bar{})
+	println(ret)
+	assert ret.contains('Response<Bar>{')
+	assert ret.contains('result: Bar{}')
+
+	ret = send(123)
+	println(ret)
+	assert ret.contains('Response<int>{')
+	assert ret.contains('result: 123')
+
+	ret = send('abc')
+	println(ret)
+	assert ret.contains('Response<string>{')
+	assert ret.contains("result: 'abc'")
+
+	ret = send(2.22)
+	println(ret)
+	assert ret.contains('Response<f64>{')
+	assert ret.contains('result: 2.22')
+}

